x/valset-pref/msg_server_test.go

```diff
@@ -124,3 +124,142 @@ func (suite *KeeperTestSuite) TestSetValidatorSetPreference() {
 		})
 	}
 }
+
+func (suite *KeeperTestSuite) TestDelegateToValidatorSet() {
+	suite.SetupTest()
+
+	amountToFund := sdk.Coins{sdk.NewInt64Coin(sdk.DefaultBondDenom, 100_000_000)} // 100 osmo
+
+	// prepare validators to delegate to
+	preferences := suite.PrepareDelegateToValidatorSet()
+
+	tests := []struct {
+		name           string
+		delegator      sdk.AccAddress
+		coin           sdk.Coin  // amount to delegate
+		expectedShares []sdk.Dec //expected shares after delegation
+		expectPass     bool
+		valSetExists   bool
+	}{
+		{
+			name:           "Delegate to valid validators",
+			delegator:      sdk.AccAddress([]byte("addr1---------------")),
+			coin:           sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(10_000_000)),
+			expectedShares: []sdk.Dec{sdk.NewDec(2_000_000), sdk.NewDec(3_320_000), sdk.NewDec(1_200_000), sdk.NewDec(3_480_000)},
+			expectPass:     true,
+		},
+		{
+			name:           "Delegate more tokens to existing validator-set",
+			delegator:      sdk.AccAddress([]byte("addr1---------------")),
+			coin:           sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(10_000_000)),
+			expectedShares: []sdk.Dec{sdk.NewDec(4_000_000), sdk.NewDec(6_640_000), sdk.NewDec(2_400_000), sdk.NewDec(6_960_000)},
+			expectPass:     true,
+			valSetExists:   true,
+		},
+		{
+			name:       "User does not have enough tokens to stake",
+			delegator:  sdk.AccAddress([]byte("addr3---------------")),
+			coin:       sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(200_000_000)),
+			expectPass: false,
+		},
+	}
+
+	for _, test := range tests {
+		suite.Run(test.name, func() {
+			// setup message server
+			msgServer := valPref.NewMsgServerImpl(suite.App.ValidatorSetPreferenceKeeper)
+			c := sdk.WrapSDKContext(suite.Ctx)
+
+			if !test.valSetExists {
+				suite.FundAcc(test.delegator, amountToFund)
+
+				_, err := msgServer.SetValidatorSetPreference(c, types.NewMsgSetValidatorSetPreference(test.delegator, preferences))
+				suite.Require().NoError(err)
+			}
+
+			// call the create validator set preference
+			_, err := msgServer.DelegateToValidatorSet(c, types.NewMsgDelegateToValidatorSet(test.delegator, test.coin))
+			if test.expectPass {
+				suite.Require().NoError(err)
+
+				// check if the user balance decreased
+				balance := suite.App.BankKeeper.GetBalance(suite.Ctx, test.delegator, sdk.DefaultBondDenom)
+				suite.Require().True(balance.IsLT(amountToFund[0]))
```

Comment:
why not check the exact amount rather than inequalities?

Reply:
Yup will do